flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/data/RowDataProjection.java

```diff
@@ -254,4 +258,78 @@ public MapData getMap(int pos) {
   public RowData getRow(int pos, int numFields) {
     return (RowData) getValue(pos);
   }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (!(o instanceof RowDataProjection)) {
+      return false;
+    }
+
+    RowDataProjection that = (RowDataProjection) o;
+    return deepEquals(that);
+  }
+
+  @Override
+  public int hashCode() {
+    int result = Objects.hashCode(getRowKind());
+    for (int pos = 0; pos < getArity(); pos++) {
+      if (!isNullAt(pos)) {
+        // Arrays.deepHashCode handles array object properly
+        result = 31 * result + Arrays.deepHashCode(new Object[] {getValue(pos)});
```

Comment:
curious, when calculating hashcode, do we need to take Map/List as a special case? Will `Arrays.deepHashCode` be able to handle that?  
Or it's because `This projection will not project the nested children types of repeated types like lists and
   maps`, so we can ignore them?

Reply:
Not so sure, but I prefer the first one.

Reply:
@yegangy0718 Java List and Map objects implements `hashCode` properly. here we are just leveraging `java.util.Arrays` to handle the hashCode for array type field.

@hililiwei what do you mean by preferring the first one?